### PR TITLE
Qt version overlay and VTK update

### PR DESCRIPTION
**Pull request by muellni**
I updated vtk to compile with GCC 6 and came up with some patches:
- make vtk and qwt link to Qt 4 using a plugin overlay `plugins/qt4`. Maybe this would resolve #1300 
- set CMake QT_QMAKE_EXECUTABLE variable to cache, because running `find_package(Qt4)` a second time in CMake somehow forgets the value of QT_QMAKE_EXECUTABLE
- fix VTK GCC 6 compilation, which I did by upgrading the Compiler Version check - this will most likely be fixed in the next VTK release, because I didn't spot the relevant code in https://github.com/Kitware/VTK/tree/master/CMake

I tested
- GCC6 Qt4`make MXE_PLUGIN_DIRS='plugins/gcc6 plugins/qt4'  MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' JOBS=8 -j4 vtk qwt`
- GCC6 Qt5`make MXE_PLUGIN_DIRS=plugins/gcc6  MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' JOBS=8 -j4 vtk qwt`
- GCC4 Qt5`make MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' JOBS=8 -j4 vtk qwt` **Failed**, because my qtbase won't build using GCC 4 with some PCH related error (logfile 
  [qtbase_i686-w64-mingw32.shared.zip](https://github.com/mxe/mxe/files/500516/qtbase_i686-w64-mingw32.shared.zip)
  )
- GCC4 Qt4 `make MXE_PLUGIN_DIRS=plugins/qt4  MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' JOBS=8 -j4 vtk qwt` **Failed**, same PCH problem (logfile 
  [qt_i686-w64-mingw32.static.zip](https://github.com/mxe/mxe/files/500548/qt_i686-w64-mingw32.static.zip))

### CODE REVIEW

**Comment by starius** [comment]
Failed to build pcl, which depends on vtk:
https://gist.github.com/53c07f77ef759c7265b2ff2c941e3360

**Comment by muellni** [comment]
PCL seems to be stuck at VTK 6. I tried to upgrade to PCL 1.8.0, but getting strange errors:
https://gist.github.com/muellni/53621f8dc81d38f27587cd00e4bbd6d8

**Comment by starius** [comment]
pcl 1.8.0 (on top of vtk 7 from this pull request) fails for me with [other error](https://gist.github.com/9f64bf8c24e754fa769110294941c096). PR https://github.com/mxe/mxe/pull/1534 fixes it.

If pcl can be updated with current vtk (it is being tested at the moment), then it should be done at first (to keep all commits of MXE buildable).

**Comment by starius** [comment]
@muellni 
I have merged #1534
Can you rebase this pull request, please?

```
git pull --rebase https://github.com/mxe/mxe master
```

**Comment by muellni** [comment]
@starius sure, thanks

**Comment by starius** [comment]
`make MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' JOBS=2 -j3 vtk qwt` works for me for commit f19edfbabf83eb309cc1b03061c2efc6bcf5fe13 in clean clone.

@muellni 
Can you reproduce failures with gcc 4?

**Comment by muellni** [comment]
@starius I Think my GCC 4 problems are unrelated to this PR. Actually they led me to creating this PR. I'm guessing Qt somehow uses my host compiler (GCC 6) for creating the PCH (which would then be incompatible to GCC 4), but I didn't investigate it yet.

**Comment by muellni** [comment]
@starius running `make MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' JOBS=2 -j3 vtk qwt` on clean f19edfbabf83eb309cc1b03061c2efc6bcf5fe13 results in same error:  https://gist.github.com/anonymous/b1517a11c5457048525202950fbb0e1d

**Comment by starius** [comment]
See also https://github.com/mxe/mxe/issues/1536
The root cause seems to be gcc 6 being used as system compiler.

**Comment by muellni** [comment]
@starius running on current master7fe43cc also fails: https://gist.github.com/a1fc31ba0f4b87b3fe4a01de7daecc2e

**Comment by starius** [comment]
gcc is being updated to 5 in https://github.com/mxe/mxe/pull/1541

Can you check if it works with gcc5 with the following command:

```
make MXE_PLUGIN_DIRS=plugins/gcc5 \
  MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' \
  JOBS=8 -j4 vtk qwt
```
?

**Comment by muellni** [comment]
Hi @starius , sorry for letting you wait so long. 
I checked out https://github.com/LuaAndC/mxe f55a93f and ran `make MXE_PLUGIN_DIRS=plugins/gcc5 \     
  MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' \
  JOBS=8 -j4 vtk qwt`. Unfortunately gcc failed to compile:
https://gist.github.com/c9088cd36c7c0d2277c8a84ae0db0654
https://gist.github.com/013ee111569679fb5df0572d89e4b1e7

**Comment by starius** [comment]
The error message is

> cfns.gperf:101:1: error: ‘const char* libc_name_p(const char*, unsigned int)’ redeclared inline with ‘gnu_inline’ attribute

It is known problem when building gcc 5.3 with gcc 6.
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=69959

I am updating gcc 5.3 to gcc 5.4 in the plugin, because the bug should be fixed in gcc 5.4.
Can you retry the command with this applied: https://github.com/mxe/mxe/pull/1556

**Comment by muellni** [comment]
Hi @starius,
I checked out f78221a861994cb1f9a932761cf1242dd0a0d97f and ran `make MXE_PLUGIN_DIRS=plugins/gcc5 \
  MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' \
  JOBS=8 -j4 vtk qwt`.
GCC built fine, but all 4 Qt variants failed:
https://gist.github.com/796a17c07af3d203971aff1f0a5df6f6
https://gist.github.com/7d4f52beafffa43c489403147f5436ad
https://gist.github.com/26fa1730ffb76cec44d46ebaf2b5873c
https://gist.github.com/f822991101e2ff5490fa615ed1edd0a4

**Comment by starius** [comment]
Try to apply #1557 as well.

If it works, check if package qtbase builds. Currently it fails with the
same error message as qt does.

27 окт. 2016 г. 9:49 пользователь "Martin Müllenhaupt" <
notifications@github.com> написал:

**Comment by muellni** [comment]
I checked out f78221a, cherry-picked b2e82794632cde9d8f0d13381f56f9b7a356e728 and ran `make MXE_PLUGIN_DIRS=plugins/gcc5 MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' JOBS=8 -j4 vtk qwt`.
Now `qt` still fails, but `qtbase` built (looks like `qt` is ignoring the `-no-pch` option):
https://gist.github.com/9455ca9293f9778caa316d12c05881eb
https://gist.github.com/f52596142ff8628b29134a7c0111a027

**Comment by starius** [comment]
I have updated https://github.com/mxe/mxe/pull/1557 (added commit "qt 4: backport patch for -no-pch from qtbase 5.7.0").

Does it work now?

**Comment by starius** [comment]
Build of qt is fixed by https://github.com/mxe/mxe/pull/1557 (works for me and in https://github.com/mxe/mxe/issues/1554#issuecomment-257027218)

This PR LGTM.

Please rebase your branch against current master:
```
git pull --rebase https://github.com/mxe/mxe master
git push -f
```

And repeat 4 tests from your original message (qt/qtbase x gcc4/gcc6).

**Comment by muellni** [comment]
```make MXE_PLUGIN_DIRS=plugins/gcc6 MXE_TARGETS='i686-w64-mingw32.shared i686-w64-mingw32.static' -j2 JOBS=4 vtk qwt``` builds

**Comment by luca-penasa** [comment]
any news on this? the vtk in mxe is starting to be quite old...

**Comment by muellni** [comment]
@starius any chance merging this PR?  Should I rebase it?

**Comment by luca-penasa** [comment]
@muellni Did you rebased that branch at the end? I am still using it cause I need vtk7 but would be cool to have it updated to the latest head... have you worked on it again?
Thanks

**Comment by muellni** [comment]
@luca-penasa I'm working on an rebased version to build VTK 8.0.0.rc1
It's currently fails after 99% compilation with `undefined reference to `vtk_haru_HPDF_Page_GRestore'` errors.

**Comment by muellni** [comment]
seems to work, but needs more build tests

**Comment by muellni** [comment]
compiling static VTK works, but linking the text executable fails with a ton of Qt related linking errors and some VTK->glew linking errors.

**Comment by muellni** [comment]
Tested:
 - `make MXE_TARGETS='i686-w64-mingw32.static' JOBS=8 -j4 vtk qwt`

**Comment by muellni** [comment]
@luca-penasa VTK 8 builds fine for Qt 5 and Qt 4, static and shared. @starius can you merge this please? Or should I refactor the commits?

**Comment by muellni** [comment]
@mabrand could you please have a look at merging this PR?

File: src/qtbase.mk [hunk 1]
```diff
@@ -102,6 +102,9 @@ define $(PKG)_BUILD
      printf 'test-qt5.exe\r\n'; \
      printf 'test-qtbase-pkgconfig.exe\r\n';) \
      > '$(PREFIX)/$(TARGET)/bin/test-qt5.bat'
+
+    # add pcre16 to CMake config of Qt5Core to fix static linking
+    $(SED) -i 's,set(_Qt5Core_LIB_DEPENDENCIES \"\"),set(_Qt5Core_LIB_DEPENDENCIES \"pcre16\"),g' '$(PREFIX)/$(TARGET)/qt5/lib/cmake/Qt5Core/Qt5CoreConfig.cmake'
```
Comment (mabrand) [inline src/qtbase.mk]: It would be better to do this by add a patch to src/qtbase-1-fixes.patch instead of a sed hack.

File: docs/packages.json [hunk 2]
```diff
@@ -172,7 +172,7 @@
     "libgpg_error": {"version": "1.27", "website": "https://www.gnupg.org/related_software/libgpg-error/", "description": "libgpg-error"},
     "libgsasl": {"version": "1.8.0", "website": "https://www.gnu.org/software/gsasl/", "description": "Libgsasl"},
     "libgsf": {"version": "1.14.30", "website": "https://developer.gnome.org/gsf/", "description": ""},
-    "libharu": {"version": "2.2.1", "website": "http://libharu.org/", "description": ""},
+    "libharu": {"version": "2.3.0", "website": "http://libharu.org/", "description": ""},
```
Comment (mabrand) [inline docs/packages.json]: This file gets updated automatically. No need for the change here.

File: docs/packages.json [hunk 3]
```diff
@@ -416,8 +416,7 @@
     "vo-aacenc": {"version": "0.1.3", "website": "https://github.com/mstorsjo/vo-aacenc", "description": "VO-AACENC"},
     "vo-amrwbenc": {"version": "0.1.3", "website": "https://github.com/mstorsjo/vo-amrwbenc", "description": "VO-AMRWBENC"},
     "vorbis": {"version": "1.3.5", "website": "http://www.vorbis.com/", "description": "Vorbis"},
-    "vtk": {"version": "5.8.0", "website": "http://www.vtk.org/", "description": ""},
-    "vtk6": {"version": "6.3.0", "website": "http://www.vtk.org/", "description": "VTK6"},
+    "vtk": {"version": "8.0.0.rc1", "website": "http://www.vtk.org/", "description": "The Visualization Toolkit"},
```
Comment (mabrand) [inline docs/packages.json]: idem

File: src/qtbase.mk [hunk 4]
```diff
@@ -103,8 +103,10 @@ define $(PKG)_BUILD
...
 
-    # add pcre16 to CMake config of Qt5Core to fix static linking
-    $(SED) -i 's,set(_Qt5Core_LIB_DEPENDENCIES \"\"),set(_Qt5Core_LIB_DEPENDENCIES \"pcre16\"),g' '$(PREFIX)/$(TARGET)/qt5/lib/cmake/Qt5Core/Qt5CoreConfig.cmake'
+    # add libs to CMake config of Qt5Core to fix static linking
+    $(SED) -i 's,set(_Qt5Core_LIB_DEPENDENCIES \"\"),set(_Qt5Core_LIB_DEPENDENCIES \"ole32;uuid;ws2_32;advapi32;shell32;user32;kernel32;mpr;version;winmm;z;pcre2-16\"),g' '$(PREFIX)/$(TARGET)/qt5/lib/cmake/Qt5Core/Qt5CoreConfig.cmake'
+    $(SED) -i 's,set(_Qt5Gui_LIB_DEPENDENCIES \"Qt5::Core\"),set(_Qt5Gui_LIB_DEPENDENCIES \"Qt5::Core;ole32;uuid;ws2_32;advapi32;shell32;user32;kernel32;mpr;version;winmm;z;pcre2-16;png16;harfbuzz;z\"),g' '$(PREFIX)/$(TARGET)/qt5/lib/cmake/Qt5Gui/Qt5GuiConfig.cmake'
+    $(SED) -i 's,set(_Qt5Widgets_LIB_DEPENDENCIES \"Qt5::Gui;Qt5::Core\"),set(_Qt5Widgets_LIB_DEPENDENCIES \"Qt5::Gui;Qt5::Core;gdi32;comdlg32;oleaut32;imm32;opengl32;png16;harfbuzz;ole32;uuid;ws2_32;advapi32;shell32;user32;kernel32;mpr;version;winmm;z;pcre2-16;shell32;uxtheme;dwmapi\"),g' '$(PREFIX)/$(TARGET)/qt5/lib/cmake/Qt5Widgets/Qt5WidgetsConfig.cmake'
 endef
```
Comment (mabrand) [inline src/qtbase.mk]: It would be better to do this by add a patch to src/qtbase-1-fixes.patch instead of a sed hack.

**Comment by muellni** [comment]
Split up into 5 PRs.

**Pull request closed without merging**